src/qtcwin.c

```diff
@@ -0,0 +1,1814 @@
+/*
+ * Tlf - contest logging program for amateur radio operators
+ * Copyright (C) 2013-2014 Ervin Hegedüs - HA2OS <airween@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+	/* ------------------------------------------------------------
+	 *        Handle QTC panel, receive and send QTC, write to log
+	 *
+	 *--------------------------------------------------------------*/
+
+#include "qtcwin.h"
+
+#include "onechar.h"
+#include <panel.h>
+#include "nicebox.h"
+#include "time_update.h"
+#include <ctype.h>
+#include "rtty.h"
+#include "sendbuf.h"
+#include "log_recv_qtc_to_disk.h"
+#include "log_sent_qtc_to_disk.h"
+#include "globalvars.h"
+#include "qtcutil.h"
+#include "genqtclist.h"
+#include "speedupndown.h"
+#include "cw_utils.h"
+#include "keyer.h"
+#include "callinput.h"
+#include "get_time.h"
+#include "write_keyer.h"
+
+#include <sys/time.h>
+
+#define DIRCLAUSE (direction == RECV) || (direction == SEND && (activefield == 0 || activefield == 2))
+
+extern char hiscall[];
+extern char lastcall[];
+extern int trxmode;
+extern t_qtcreclist qtcreclist;
+extern t_qtclist qtclist;
+extern int keyerport;
+extern int nr_qsos;
+extern char qtc_recv_msgs[12][80];
+extern char qtc_send_msgs[12][80];
+extern char qtc_phrecv_message[14][80];
+extern char qtc_phsend_message[14][80];
+extern int data_ready;
+extern struct t_qtc_store_obj *qtc_temp_obj;
+extern int qtcrec_record;
+extern char qtcrec_record_command[2][50];
+extern char qtcrec_record_command_shutdown[50];
+extern char wkeyerbuffer[];
+extern t_qtc_ry_line qtc_ry_lines[QTC_RY_LINE_NR];
+extern int qtc_ry_currline;
+extern int qtc_ry_capture;
+
+enum {
+  QTCRECVWINBG = 32,
+  QTCRECVLINE,
+  QTCRECVINVLINE,
+  QTCRECVCURRLINE,
+  QTCRECVBG,
+  KEYERLINE
+};
+
+char sentqtc_queue[11][30];
+int sentqtc_p;
+
+int qtcpanel = 0;
+int currrecstate;
+WINDOW * qtcwin;
+PANEL * qtc_panel;
+WINDOW * ry_win;
+PANEL * ry_panel;
+WINDOW * ry_help_win;
+PANEL * ry_help_panel;
+int activefield;
+// array values: hor position, cursor position, field len to fill with spaces
+int pos[6][3] = {{3, 3, 15}, {3, 6, 4}, {8, 8, 2}, {3, 3, 4}, {8, 8, 14}, {24, 24, 4}};
+int curpos = 0;
```

Comment:
'Curpos' stores the current position, but not (just) from the and to the begin, in backwards direction. Sometimes it counts as backward, somtimes as forward. Here is a snippet for all cases:
```C
	  case 155:	// left
		    if (DIRCLAUSE) {
			if (curpos < curfieldlen) {	// curpos is a shift, means lenght - position
			    curpos++;
			    showfield(activefield);
			}
		    }
		    break;
	  case 154:	// right
		    if (DIRCLAUSE) {
			if (curpos > 0) {
			    curpos--;
			    showfield(activefield);
			}
		    }
		    break;
```